pkg/suggestion/v1beta1/pbt/service.py

```diff
@@ -0,0 +1,392 @@
+            logger.warning(
+                "Trial generation >0 but no previous checkpoint trial provided"
+            )
+        obj = PbtJob(assignments, generation, parent)
+        self.pending.append(obj)
+        new_trial_dir = os.path.join(_DATA_PATH, obj.uid)
+        if parent is None:
+            os.makedirs(new_trial_dir, exist_ok=True)
+        else:
+            previous_trial_dir = os.path.join(_DATA_PATH, obj.parent)
+            shutil.copytree(previous_trial_dir, new_trial_dir)
+
+        logger.info("PbtJob enqueued: {}".format(obj))
+        return obj.uid
+
+    def get(self):
+        if len(self.pending) == 0:
+            raise RuntimeError("Pending queue is empty!")
+
+        # Move job to running queue
+        obj = self.pending.pop(0)
+        logger.info("PbtJob submitted: {}".format(obj))
+        self.running[obj.uid] = obj
+
+        return obj.get()
+
+    def verify_running(self, trials):
+        nospawn_uids = list(self.running.keys())
+        for trial in trials:
+            trial_annotations = trial.spec.annotations
+            uid = trial_annotations["pbt.suggestion.katib.kubeflow.org/uid"]
+            if uid in nospawn_uids:
+                nospawn_uids.remove(uid)
+        for uid in nospawn_uids:
+            logger.info("PbtJob requeued: {}".format(uid))
+            self.pending.append(self.running.pop(uid))
+
+    def update(self, trial):
+        trial_assignments = [Assignment.convert(assignment) for assignment in trial.spec.parameter_assignments.assignments]
+        trial_annotations = trial.spec.annotations
+        uid = trial_annotations["pbt.suggestion.katib.kubeflow.org/uid"]
+        generation = trial_annotations["pbt.suggestion.katib.kubeflow.org/generation"]
+
+        # Do not update active/pending or already processed trials
+        if trial.status.condition in (
+            api_pb2.TrialStatus.TrialConditionType.CREATED,
+            api_pb2.TrialStatus.TrialConditionType.RUNNING,
+        ):
+            return
+        for i in range(self.generation + 1):
+            if uid in self.completed[i]:
+                return
+
+        if not uid in self.running:
+            logger.error("Unable to find trial match in PbtJobQueue.running!")
+            return
+        obj = self.running.pop(uid)
+        obj.metric_value = self._get_objective_value(trial)
+        self.completed[self.generation][obj.uid] = obj
+
+        # Generate next trial if necessary
+        if (
+            trial.status.condition
+            in (
+                api_pb2.TrialStatus.TrialConditionType.SUCCEEDED,
+                api_pb2.TrialStatus.TrialConditionType.EARLYSTOPPED,
+            )
+            and Trial.convertTrial(trial) is None
+        ):
+            # Trial invalid, retry
+            logger.error(
+                "Unable to convert trial {} (status: {}), re-adding to task queue".format(
+                    trial.name, trial.status.condition
+                )
+            )
```

Comment:
I think you can avoid this use-case, since it's the Controller responsibility that all Succeeded and EarlyStopped Trials have reported metrics.
WDYT @a9p @johnugeorge @gaocegege ?